hazelcast-sql/src/main/java/com/hazelcast/sql/impl/type/QueryDataType.java

```diff
@@ -325,17 +383,54 @@ public boolean equals(Object o) {
             return false;
         }
         QueryDataType that = (QueryDataType) o;
-        return !isCustomType()
-                ? converter.getId() == that.converter.getId()
-                : objectTypeName.equals(that.objectTypeName)
-                        && objectTypeKind == that.objectTypeKind
-                        && Objects.equals(objectTypeMetadata, that.objectTypeMetadata)
-                        && objectFields.equals(that.objectFields);
+        return !isCustomType() ? converter == that.converter : getDigest().equals(that.getDigest());
```

Comment:
is it possible to create UDT with special characters in type name or field names (by escaping)? If so, comparing digest could be unreliable

Reply:
8cf82f68

Reply:
are you sure we need escaping? It also might be inconsistent with standard SQL escaping (not that it matters much).

From what I have seen Calcite does not care about escaping, but usually does not have digests with user-supplied names. Another approach I have seen there is to have prefix in the digest. See eg. `RelRecordType.generateTypeString` which generates `RecordType(<fields>)`

Reply:
Unless we escape metacharacters, users can develop attacks for our digest comparison, regardless of whether we use a prefix. I didn't understand how standard SQL escaping is relevant to us.

Reply:
check for example `new RelRecordType(List.of(new RelDataTypeFieldImpl("aaa, RecordType(INTEGER(16) NOT NULL aaa) NOT NULL", 0, HazelcastIntegerType.create(16, false)))).getFullTypeString()`

Reply:
What does this prove?

Reply:
that Calcite does not escape so maybe we also don't have to?

Reply:
I don't understand why/how Calcite limits us with this respect. Calcite has a very serious security flaw, let's introduce it to Hazelcast as well to be consistent. It just does not make sense to me.

Reply:
✔️

Reply:
The digest string is currently not exposed to users (via `toString()` for example), so consistency with SQL is not a concern for now. We might want to reset `DATATYPE_CACHE` in each query execution since no state is supposed to keep across executions. So, it doesn't only pose a security risk, but also violates our assumption that query execution is stateless.

Reply:
> We might want to reset DATATYPE_CACHE in each query execution

probably a bad idea because queries can be parsed concurrently